dell/src/main/java/org/apache/iceberg/dell/ecs/EcsCatalog.java

```diff
@@ -0,0 +1,520 @@
+import java.io.InputStream;
+import java.io.UncheckedIOException;
+import java.util.Arrays;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Consumer;
+import java.util.stream.Collectors;
+import org.apache.iceberg.BaseMetastoreCatalog;
+import org.apache.iceberg.CatalogProperties;
+import org.apache.iceberg.CatalogUtil;
+import org.apache.iceberg.TableMetadata;
+import org.apache.iceberg.TableOperations;
+import org.apache.iceberg.catalog.Namespace;
+import org.apache.iceberg.catalog.SupportsNamespaces;
+import org.apache.iceberg.catalog.TableIdentifier;
+import org.apache.iceberg.dell.DellClientFactories;
+import org.apache.iceberg.dell.DellProperties;
+import org.apache.iceberg.exceptions.AlreadyExistsException;
+import org.apache.iceberg.exceptions.NamespaceNotEmptyException;
+import org.apache.iceberg.exceptions.NoSuchNamespaceException;
+import org.apache.iceberg.exceptions.NoSuchTableException;
+import org.apache.iceberg.hadoop.Configurable;
+import org.apache.iceberg.io.FileIO;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class EcsCatalog extends BaseMetastoreCatalog
+    implements Closeable, SupportsNamespaces, Configurable<Object> {
+
+  /**
+   * Suffix of table metadata object
+   */
+  private static final String TABLE_OBJECT_SUFFIX = ".table";
+
+  /**
+   * Suffix of namespace metadata object
+   */
+  private static final String NAMESPACE_OBJECT_SUFFIX = ".namespace";
+
+  /**
+   * Key of properties version in ECS object user metadata.
+   */
+  private static final String PROPERTIES_VERSION_USER_METADATA_KEY = "iceberg_properties_version";
+
+  private static final Logger LOG = LoggerFactory.getLogger(EcsCatalog.class);
+
+  private S3Client client;
+  private Object hadoopConf;
+  private String catalogName;
+  /**
+   * Warehouse is unified with other catalog that without delimiter.
+   */
+  private String warehouseLocation;
+  private DellProperties dellProperties;
+  private PropertiesSerDes propertiesSerDes;
+  private FileIO fileIO;
+
+  /**
+   * No-arg constructor to load the catalog dynamically.
+   * <p>
+   * All fields are initialized by calling {@link EcsCatalog#initialize(String, Map)} later.
+   */
+  public EcsCatalog() {
+  }
+
+  @Override
+  public void initialize(String name, Map<String, String> properties) {
+    this.catalogName = name;
+    this.dellProperties = new DellProperties(properties);
+    this.warehouseLocation =
+        cleanWarehouse(properties.get(CatalogProperties.WAREHOUSE_LOCATION), dellProperties.ecsCatalogDelimiter());
+    this.client = DellClientFactories.from(properties).ecsS3();
+    this.propertiesSerDes = PropertiesSerDes.current();
+    this.fileIO = initializeFileIO(properties);
```

Comment:
Seems we've just missed to close the `fileIO`, right ?

Reply:
Add CloseableGroup to close `fileIO`.